packager/mpd/base/xml/xml_node.cc

```diff
@@ -56,8 +57,16 @@ bool IsTimelineConstantDuration(const std::list<SegmentInfo>& segment_infos,
     return false;
 
   const SegmentInfo& first_segment = segment_infos.front();
-  if (first_segment.start_time != first_segment.duration * (start_number - 1))
-    return false;
+
+  if (first_segment.duration < target_duration) {
+    if (static_cast<uint32_t>(first_segment.start_time / target_duration) !=
+        start_number)
+      return false;
+  } else {
+    if (static_cast<uint32_t>(first_segment.start_time /
+                              first_segment.duration) != start_number - 1)
+      return false;
+  }
```

Comment:
The logic is incorrect. We'll need to handle the possible short segment here if we do not want to remove it earlier. There are a few problems with the code above. For example, the code will return false in line 76 when there is a shorter initial segment.

We can handle it properly with:

```c++
bool GetTimelineConstantDuration(const std::list<SegmentInfo>& segment_infos,
                                 uint32_t start_number,
                                 int64_t* constant_duration) {
  if (!FLAGS_segment_template_constant_duration)
    return false;

  DCHECK(!segment_infos.empty());
  if (segment_infos.size() > 2)
    return false;

  int64_t first_segment_timestamp = segment_infos.front().start_time;

  if (segment_infos.size() == 1) {
    *constant_duration = segment_infos.front().duration; 
    return first_segment_timestamp / *constant_duration == start_number - 1;
  }

  if (segment_infos.front().repeat != 0 && segment_infos.back().repeat() != 0)
    return false;

  int64_t last_segment_timestamp = segment_infos.back().start_time;
  int64_t last_segment_start_number = start_number + segment_infos.front().repeat + 1;
  *constant_duration = 
      segment_infos.front().repeat == 0 ? segment_infos.back().duration :  // shorter initial segment
                                          segment_infos.front().duration;  // shorter ending segment
  return first_segment_timestamp / *constant_duration == start_number - 1 &&
         last_segment_timestamp / *constant_duration == last_segment_start_number - 1;    
}
```

Reply:
After adding 
```
  if (segment_infos.size() == 1) {
    *constant_duration = segment_infos.front().duration;
    return first_segment_timestamp / *constant_duration == start_number - 1;
  }
```
and the first segment is shorter than target segment, then won't the calculated start time in representation.cc (start_time / scaled_target_duration + 1) be different than the one that it is compared against here? Also if we calculate the start number based on smaller duration than it will be greater than one calculated in   ChunkingHandler::OnMediaSample and representation.cc.

`For example, the code will return false in line 76 when there is a shorter initial segment.
`
As per my understanding: 
1. If smaller segment is added, then IsTimelineConstantDuration will return true at 72.  
2. If there is a target duration segment added, then  IsTimelineConstantDuration will return true at 82.
3. Then as soon as the target duration repeat count goes above 0, the code will pop out the first shorter segment and in this case, the method will return true at line 72 and won't reach 76. 

Can you please clarify?

Reply:
> Then as soon as the target duration repeat count goes above 0, the code will pop out the first shorter segment and in this case, the method will return true at line 72 and won't reach 76.

The initial shorter segment will not be removed until the live window is filled up.

Let's assume a buffer depth of 10000 and target duration of 1000.
(1) Initially there is only the initial shorter segment: S1 (timestamp, duration) where timestamp = 10500, duration = 500
(2) S2 (11000, 1000)
(3) S3 (12000, 1000)
(4) S4 (13000, 1000)
...
(10) S10 (19000, 1000)
(11) S11 (20000, 1000)
(12) S12 (21000, 1000)
(13) S12 (22000, 1000)

The start number is 10500 / 1000 + 1 = 11.

segment_infos is
(1) [(timestamp, duration, repeat)], where timestamp = 10500, duration = 500, repeat = 0
(2) [(10500, 500, 0), (11000, 1000, 0)]
(3) [(10500, 500, 0), (11000, 1000, 1)]
(4) [(10500, 500, 0), (11000, 1000, 2)]
...
(10) [(10500, 500, 0), (11000, 1000, 8)]
(11) [(10500, 500, 0), (11000, 1000, 9)]
(12) [(11000, 1000, 10)]
(13) [(12000, 1000, 10)]

With your code, 
(1) True, constant_duration = 1000
(2) True, constant_duration = 1000
(3) False
(4) False
...
(10) False
(11) False
(12) True, constant_duration = 1000
(13) True, constant_duration = 1000

With my proposed code, 
(1) False
(2) True, constant_duration = 1000
(3) True, constant_duration = 1000
(4) True, constant_duration = 1000
...
(10) True, constant_duration = 1000
(11) True, constant_duration = 1000
(12) True, constant_duration = 1000
(13) True, constant_duration = 1000

You are right that (1) fails in my proposed code, which I intentionally left out as I don't want the code to be dependent on another parameter. If you feel strongly about it, we can also merge your solution and my solution so it returns true in all cases.

Reply:
Thanks for the example. By popping out, I meant code added at representation.cc::368 to 380 to pop the first short segment if approximate_segment_timeline and target_segment_duration are set. At (3), the first shorter segment will be popped out and  only [(11000, 1000, 1] will be present and True will be returned from line 72 in xml_node.cc.

After trying out the changes with flags set to --mpd_output live.mpd --allow_approximate_segment_timeline --segment_template_constant_duration --segment_duration 10 --time_shift_buffer_depth 50, I get the following values of segment_infos_: 
```
[0413/204712:ERROR:representation.cc(172)] SEGMENT_INFO List: 
[0413/204712:ERROR:representation.cc(175)] 0  Start Time: 1798126434  Duration: 180000  Repeat: 0  -> shorter 2 seconds segment
[0413/204722:ERROR:representation.cc(172)] SEGMENT_INFO List: 
[0413/204722:ERROR:representation.cc(175)] 0  Start Time: 1798126434  Duration: 180000  Repeat: 0
[0413/204722:ERROR:representation.cc(175)] 1  Start Time: 1798306434  Duration: 900000  Repeat: 0 

[0413/204732:ERROR:representation.cc(172)] SEGMENT_INFO List: 
[0413/204732:ERROR:representation.cc(175)] 0  Start Time: 1798306434  Duration: 900000  Repeat: 1     -> shorter segment popped out

[0413/204742:ERROR:representation.cc(172)] SEGMENT_INFO List: 
[0413/204742:ERROR:representation.cc(175)] 0  Start Time: 1798306434  Duration: 900000  Repeat: 2

[0413/204752:ERROR:representation.cc(172)] SEGMENT_INFO List: 
[0413/204752:ERROR:representation.cc(175)] 0  Start Time: 1798306434  Duration: 900000  Repeat: 3

[0413/204802:ERROR:representation.cc(172)] SEGMENT_INFO List: 
[0413/204802:ERROR:representation.cc(175)] 0  Start Time: 1798306434  Duration: 900000  Repeat: 4

[0413/204812:ERROR:representation.cc(172)] SEGMENT_INFO List: 
[0413/204812:ERROR:representation.cc(175)] 0  Start Time: 1798306434  Duration: 900000  Repeat: 5

[0413/204822:ERROR:representation.cc(172)] SEGMENT_INFO List: 
[0413/204822:ERROR:representation.cc(175)] 0  Start Time: 1799206434  Duration: 900000  Repeat: 5 

[0413/204832:ERROR:representation.cc(172)] SEGMENT_INFO List: 
[0413/204832:ERROR:representation.cc(175)] 0  Start Time: 1800106434  Duration: 900000  Repeat: 5 -> repeat count remains 5 due to tsbd and start time increases.

```
Is this correct? What am I missing?

Reply:
Ah, I completely missed your change representation.cc::368 to 380. Sorry about that. Yes, I think that works. How does the generated mpd looks like with this change using your testing live stream? Are both audio and video representations the same as expected?

Reply:
The generated mpd for video only looks like: 
```
<?xml version="1.0" encoding="UTF-8"?>
<!--Generated with https://github.com/google/shaka-packager version ce932f68a2-debug
-->
<MPD xmlns="urn:mpeg:dash:schema:mpd:2011" xmlns:xsi="http://www.w3.org/2001/XMLSche
ma-instance" xsi:schemaLocation="urn:mpeg:dash:schema:mpd:2011 DASH-MPD.xsd" profile
s="urn:mpeg:dash:profile:isoff-live:2011" minBufferTime="PT2S" type="dynamic" publis
hTime="2020-04-14T21:17:12Z" availabilityStartTime="2020-04-13T22:14:20Z" minimumUpd
atePeriod="PT5S" timeShiftBufferDepth="PT50S">
  <Period id="0" start="PT0S">
    <AdaptationSet id="0" contentType="video" width="1920" height="1080" frameRate="
90000/3750" segmentAlignment="true" par="16:9">
      <Representation id="0" bandwidth="2991070" codecs="avc1.4d401f" mimeType="vide
o/mp4" sar="1:1">
        <SegmentTemplate timescale="90000" initialization="live-video.mp4" media="li
ve-video-$Number$.mp4" startNumber="7822" duration="900000"/>
      </Representation>
    </AdaptationSet>
  </Period>
</MPD>
```
Trying to test muxed audio and video ts file and I do not see a separate audio and video adaptation set. How to get adaptation sets for audio and video? I am using the following command: 
`packager 'input=udp://127.0.0.1:1234?interface=0.0.0.0,stream=video,init_segment=live-video.mp4,segment_template=live-video-$Number$.mp4' --mpd_output live.mpd --allow_approximate_segment_timeline --segment_template_constant_duration --segment_duration 10 --time_shift_buffer_depth 50`

Reply:
You can add another stream descriptor for audio, e.g.

```shell
packager 'input=udp://127.0.0.1:1234?interface=0.0.0.0,stream=video,init_segment=live-video.mp4,segment_template=live-video-$Number$.mp4' 'input=udp://127.0.0.1:1234?interface=0.0.0.0,stream=audio,init_segment=live-audio.mp4,segment_template=live-audio-$Number$.mp4'  --mpd_output live.mpd --allow_approximate_segment_timeline --segment_template_constant_duration --segment_duration 10 --time_shift_buffer_depth 50
```

Reply:
Following are the mpds generated with audio and video adaptation sets: 
```
<?xml version="1.0" encoding="UTF-8"?>
<!--Generated with https://github.com/google/shaka-packager version ce932f68a2-debug-->
<MPD xmlns="urn:mpeg:dash:schema:mpd:2011" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="urn:mpeg:dash:schema:mpd:2011 DASH-MPD.xsd" p
rofiles="urn:mpeg:dash:profile:isoff-live:2011" minBufferTime="PT2S" type="dynamic" publishTime="2020-04-16T04:46:43Z" availabilityStartTime="2020-04-15T19:15:35Z
" minimumUpdatePeriod="PT5S" timeShiftBufferDepth="PT50S">
  <Period id="0" start="PT0S">
    <AdaptationSet id="0" contentType="audio" segmentAlignment="true">
      <Representation id="0" bandwidth="130708" codecs="mp4a.40.2" mimeType="audio/mp4" audioSamplingRate="48000">
        <AudioChannelConfiguration schemeIdUri="urn:mpeg:dash:23003:3:audio_channel_configuration:2011" value="2"/>
        <SegmentTemplate timescale="90000" initialization="live-audio.mp4" media="live-audio-$Number$.mp4" startNumber="3420" duration="900000"/>
      </Representation>
    </AdaptationSet>
    <AdaptationSet id="1" contentType="video" width="1920" height="1080" frameRate="90000/3750" segmentAlignment="true" par="16:9">
      <Representation id="1" bandwidth="2758145" codecs="avc1.4d401f" mimeType="video/mp4" sar="1:1">
        <SegmentTemplate timescale="90000" initialization="live-video.mp4" media="live-video-$Number$.mp4" startNumber="3422" duration="900000"/>
      </Representation>
    </AdaptationSet>
  </Period>
</MPD>
```

and progresses towards: 
```
<?xml version="1.0" encoding="UTF-8"?>
<!--Generated with https://github.com/google/shaka-packager version ce932f68a2-debug-->
<MPD xmlns="urn:mpeg:dash:schema:mpd:2011" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="urn:mpeg:dash:schema:mpd:2011 DASH-MPD.xsd" p
rofiles="urn:mpeg:dash:profile:isoff-live:2011" minBufferTime="PT2S" type="dynamic" publishTime="2020-04-16T04:47:33Z" availabilityStartTime="2020-04-15T19:15:35Z
" minimumUpdatePeriod="PT5S" timeShiftBufferDepth="PT50S">
  <Period id="0" start="PT0S">
    <AdaptationSet id="0" contentType="audio" segmentAlignment="true">
      <Representation id="0" bandwidth="130815" codecs="mp4a.40.2" mimeType="audio/mp4" audioSamplingRate="48000">
        <AudioChannelConfiguration schemeIdUri="urn:mpeg:dash:23003:3:audio_channel_configuration:2011" value="2"/>
        <SegmentTemplate timescale="90000" initialization="live-audio.mp4" media="live-audio-$Number$.mp4" startNumber="3425">
          <SegmentTimeline>
            <S t="3083403612" d="896640"/>
            <S t="3084300252" d="900480" r="4"/>
          </SegmentTimeline>
        </SegmentTemplate>
      </Representation>
    </AdaptationSet>
    <AdaptationSet id="1" contentType="video" width="1920" height="1080" frameRate="90000/3750" segmentAlignment="true" par="16:9">
      <Representation id="1" bandwidth="2990305" codecs="avc1.4d401f" mimeType="video/mp4" sar="1:1">
        <SegmentTemplate timescale="90000" initialization="live-video.mp4" media="live-video-$Number$.mp4" startNumber="3427" duration="900000"/>
      </Representation>
    </AdaptationSet>
  </Period>
</MPD>
```
and 
```
<?xml version="1.0" encoding="UTF-8"?>
<!--Generated with https://github.com/google/shaka-packager version ce932f68a2-debug-->
<MPD xmlns="urn:mpeg:dash:schema:mpd:2011" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instan
ce" xsi:schemaLocation="urn:mpeg:dash:schema:mpd:2011 DASH-MPD.xsd" profiles="urn:mpeg:dash:p
rofile:isoff-live:2011" minBufferTime="PT2S" type="dynamic" publishTime="2020-04-16T04:55:53Z
" availabilityStartTime="2020-04-15T19:15:35Z" minimumUpdatePeriod="PT5S" timeShiftBufferDept
h="PT50S">
  <Period id="0" start="PT0S">
    <AdaptationSet id="0" contentType="audio" segmentAlignment="true">
      <Representation id="0" bandwidth="130815" codecs="mp4a.40.2" mimeType="audio/mp4" audio
SamplingRate="48000">
        <AudioChannelConfiguration schemeIdUri="urn:mpeg:dash:23003:3:audio_channel_configura
tion:2011" value="2"/>
        <SegmentTemplate timescale="90000" initialization="live-audio.mp4" media="live-audio-
$Number$.mp4" startNumber="3475">
          <SegmentTimeline>
            <S t="3128400990" d="900000" r="5"/>
          </SegmentTimeline>
        </SegmentTemplate>
      </Representation>
    </AdaptationSet>
    <AdaptationSet id="1" contentType="video" width="1920" height="1080" frameRate="90000/375
0" segmentAlignment="true" par="16:9">
      <Representation id="1" bandwidth="2990305" codecs="avc1.4d401f" mimeType="video/mp4" sa
r="1:1">
        <SegmentTemplate timescale="90000" initialization="live-video.mp4" media="live-video-
$Number$.mp4" startNumber="3477" duration="900000"/>
      </Representation>
    </AdaptationSet>
  </Period>
</MPD>

```

Reply:
I think we were too conservative when we check if the timestamp is equal: https://github.com/google/shaka-packager/blob/master/packager/mpd/base/representation.cc#L383. 

I think we can increase the error threshold. Can you try what happens if we increase the error threshold to 0.1 in https://github.com/google/shaka-packager/blob/master/packager/mpd/base/representation.cc#L395?

Reply:
@kqyang Thanks. Looks like the issue was with the source file. After adding  -async 1 option while creating mpeg2ts from mp4,  I am get the following mpd:

```
<?xml version="1.0" encoding="UTF-8"?>
<!--Generated with https://github.com/google/shaka-packager version ce932f68a2-debug-->
<MPD xmlns="urn:mpeg:dash:schema:mpd:2011" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="urn:mpeg:dash:schema:mpd:2011 DASH-
MPD.xsd" profiles="urn:mpeg:dash:profile:isoff-live:2011" minBufferTime="PT2S" type="dynamic" publishTime="2020-04-18T02:47:11Z" availabilityStartTime="
2020-04-18T01:48:38Z" minimumUpdatePeriod="PT5S" timeShiftBufferDepth="PT50S">
  <Period id="0" start="PT0S">
    <AdaptationSet id="0" contentType="audio" segmentAlignment="true">
      <Representation id="0" bandwidth="130986" codecs="mp4a.40.2" mimeType="audio/mp4" audioSamplingRate="48000">
        <AudioChannelConfiguration schemeIdUri="urn:mpeg:dash:23003:3:audio_channel_configuration:2011" value="2"/>
        <SegmentTemplate timescale="90000" initialization="live-audio.mp4" media="live-audio-$Number$.mp4" startNumber="347" duration="900000"/>
      </Representation>
    </AdaptationSet>
    <AdaptationSet id="1" contentType="video" width="1920" height="1080" frameRate="90000/3750" segmentAlignment="true" par="16:9">
      <Representation id="1" bandwidth="2990276" codecs="avc1.4d401f" mimeType="video/mp4" sar="1:1">
        <SegmentTemplate timescale="90000" initialization="live-video.mp4" media="live-video-$Number$.mp4" startNumber="347" duration="900000"/>
      </Representation>
    </AdaptationSet>
  </Period>
</MPD>
 
```
Tired increasing the error threshold but the video frame duration is 3750, audio frame duration is 1920 and time scale is 90000. As 
```
std::min(frame_duration_,                                                
                static_cast<uint32_t>(kErrorThresholdSeconds *                  
                                     media_info_.reference_time_scale()));
```
looks like frame duration will always be selected as error threshold.

Reply:
> Looks like the issue was with the source file. After adding -async 1 option while creating mpeg2ts from mp4, I am get the following mpd:

Great! This looks like exactly what we want.

> looks like frame duration will always be selected as error threshold.

If needed, I am fine with removing the frame duration from the calculation.

Back to the original comment,  I am still not very comfortable with depending on an additional parameter in IsTimelineConstantDuration. 

Also, with the shorter segment removed, it isn't really necessary to have the additional parameter. 

We can keep the original code other than changing 

```c++
  if (first_segment.start_time != first_segment.duration * (start_number - 1))
```

to 

```c++
  if (first_segment.start_time / first_segment.duration != * (start_number - 1))
```

It may return false for the very first two segments, but that is fine, as it is true that the very first segment is not really compliant to the constant duration.

Reply:
I have made the changes as per above comments.
> If needed, I am fine with removing the frame duration from the calculation.
Sounds good. We can remove it if the problem still persists. What can be the limit of the error threshold?

Reply:
Further, looking at sidx boxes in the segments generated, the subsegment_duration look consistent for both video and audio and within error threshold but sometimes seem to go way over the error threshold for audio (example: 935040).  
```
Video subsegment duration: 
live-video-100.mp4
  entry 0000 = reference_type=0, referenced_size=2827855, subsegment_duration=900007, starts_with_SAP=1, SAP_type=1, SAP_delta_time=0
live-video-101.mp4
  entry 0000 = reference_type=0, referenced_size=2997414, subsegment_duration=900007, starts_with_SAP=1, SAP_type=1, SAP_delta_time=0
live-video-102.mp4
  entry 0000 = reference_type=0, referenced_size=2714349, subsegment_duration=900007, starts_with_SAP=1, SAP_type=1, SAP_delta_time=0
live-video-103.mp4
  entry 0000 = reference_type=0, referenced_size=2573218, subsegment_duration=900008, starts_with_SAP=1, SAP_type=1, SAP_delta_time=0
live-video-104.mp4

Audio subsegement_duration: 
 
entry 0000 = reference_type=0, referenced_size=163055, subsegment_duration=900480, starts_with_SAP=1, SAP_type=1, SAP_delta_time=0
live-audio-101.mp4
  entry 0000 = reference_type=0, referenced_size=162631, subsegment_duration=898560, starts_with_SAP=1, SAP_type=1, SAP_delta_time=0
live-audio-102.mp4
  entry 0000 = reference_type=0, referenced_size=162238, subsegment_duration=900480, starts_with_SAP=1, SAP_type=1, SAP_delta_time=0
live-audio-103.mp4
  entry 0000 = reference_type=0, referenced_size=163030, subsegment_duration=900480, starts_with_SAP=1, SAP_type=1, SAP_delta_time=0
live-audio-104.mp4
```

On player side,
Exoplayer sidx parser calculates the duration based on  segment_duration: 
https://github.com/google/ExoPlayer/blob/7d3f54a375fac04a746ca76a8f2e1ad32c8b45b2/library/core/src/main/java/com/google/android/exoplayer2/extractor/mp4/FragmentedMp4Extractor.java#L1155

Segment Duration is retrieved at:
 https://github.com/google/ExoPlayer/blob/7d3f54a375fac04a746ca76a8f2e1ad32c8b45b2/library/dash/src/main/java/com/google/android/exoplayer2/source/dash/manifest/SegmentBase.java#L183
while loading media chunks at the given renderer position and in our case it should rely on :
`((duration * C.MICROS_PER_SECOND) / timescale). `
@kqyang Is my understanding correct or am I missing anything? Thanks.